mayastor/src/target/iscsi.rs

```diff
@@ -83,100 +89,104 @@ thread_local! {
+        return Err(e);
     }
-    debug!("Created default iscsi portal group");
-
-    unsafe {
-        if spdk_iscsi_init_grp_create_from_initiator_list(
-            0,
-            1,
-            &mut (initiator_host.as_ptr() as *mut c_char) as *mut _,
-            1,
-            &mut (initiator_netmask.as_ptr() as *mut c_char) as *mut _,
-        ) != 0
-        {
-            spdk_iscsi_portal_grp_release(pg);
-            return Err(Error::CreateInitiatorGroup {});
-        }
+    if let Err(e) = create_initiator_group(ISCSI_INITIATOR_GROUP) {
+        destroy_portal_group(ISCSI_PORTAL_GROUP_BE);
+        destroy_portal_group(ISCSI_PORTAL_GROUP_FE);
+        return Err(e);
     }
     ADDRESS.with(move |addr| {
         *addr.borrow_mut() = Some(address.to_owned());
     });
-    debug!("Created default iscsi initiator group");
+    debug!("Created default iscsi initiator group and portal groups for address {}", address);
 
     Ok(())
 }
 
-/// Destroy iscsi default portal and initiator group.
+/// Destroy iscsi portal and initiator groups.
+fn destroy_iscsi_groups() {
+    destroy_initiator_group(ISCSI_INITIATOR_GROUP);
+    destroy_portal_group(ISCSI_PORTAL_GROUP_FE);
+    destroy_portal_group(ISCSI_PORTAL_GROUP_BE);
+}
+
 pub fn fini() {
-    unsafe {
-        let ig = spdk_iscsi_init_grp_unregister(0);
-        if !ig.is_null() {
-            spdk_iscsi_init_grp_destroy(ig);
-        }
-        let pg = spdk_iscsi_portal_grp_unregister(0);
-        if !pg.is_null() {
-            spdk_iscsi_portal_grp_release(pg);
-        }
-    }
+    destroy_iscsi_groups();
 }
 
 /// Export given bdev over iscsi. That involves creating iscsi target and
 /// adding the bdev as LUN to it.
-pub fn share(uuid: &str, bdev: &Bdev) -> Result<()> {
+pub fn share(uuid: &str, bdev: &Bdev, side: Side) -> Result<()> {
+
+    let iqn = match side {
+        Side::FrontEnd => construct_iscsi_target(uuid, bdev, ISCSI_PORTAL_GROUP_FE, ISCSI_INITIATOR_GROUP)?,
+        Side::BackEnd => construct_iscsi_target(uuid, bdev, ISCSI_PORTAL_GROUP_BE, ISCSI_INITIATOR_GROUP)?,
+    };
+    info!("Created iscsi target {} for {}", iqn, uuid );
+    Ok(())
+}
+
+/// Undo export of a bdev over iscsi done above.
+pub async fn unshare(uuid: &str) -> Result<()> {
+    let (sender, receiver) = oneshot::channel::<ErrnoResult<()>>();
     let iqn = target_name(uuid);
     let c_iqn = CString::new(iqn.clone()).unwrap();
-    let mut group_idx: c_int = 0;
+
+    info!("Destroying iscsi target {}", iqn);
+
+    unsafe {
+        spdk_iscsi_shutdown_tgt_node_by_name(
+            c_iqn.as_ptr(),
+            Some(done_errno_cb),
+            cb_arg(sender),
+        );
+    }
+    receiver
+        .await
+        .expect("Cancellation is not supported")
+        .context(DestroyTarget {})?;
+    info!("Destroyed iscsi target {}", uuid);
+    Ok(())
+}
+
+fn construct_iscsi_target(bdev_name: &str, bdev: &Bdev, pg_idx: c_int, ig_idx: c_int ) -> Result<String ,Error>{
```

Comment:
For readability I would rename the function as share_... rather than construct...
Also for readability the first argument should be named uuid instead of bdev_name.
Finally for easier to understand diff output I would move this function ahead of share function.

Reply:
I'll change the function name.
Actually, I'll change the function parameter name to bdev_name throughout. It is no longer only a uuid but when called from nexus share it can have the prefix "nexus-"  and also "crypto-" if enabled. (I should also have changed the parameter name in share() and unshare() when I recently changed the code to go through those functions in both cases but neglected to do so.)
I'll move the function up, as suggested.